src/sagemaker/session.py

```diff
@@ -855,6 +856,12 @@ def train(  # noqa: C901
         inferred_resource_config = update_nested_dictionary_with_values_from_config(
             resource_config, TRAINING_JOB_RESOURCE_CONFIG_PATH, sagemaker_session=self
         )
+        if (
+            "InstanceType" in inferred_resource_config
+            and not instance_supports_kms(inferred_resource_config["InstanceType"])
+            and "VolumeKmsKeyId" in inferred_resource_config
+        ):
+            del inferred_resource_config["VolumeKmsKeyId"]
```

Comment:
These instances don't have any notion of KMS, so there's no security problem, AFAIK. If *any* instance though supports KMS, we always apply the KMS key.

Reply:
The admin sets the KMS config with the expectation it is used whenever KMS is supported. But if someone uses a instance type like `local` mode or a g5 instance, the KMS settings makes no sense, and shouldn't be applied.

Reply:
Let me add some more context here.

The only thing you can encrypt with a custom KMS key is an EBS volume. The issue here is that you cannot attach an EBS volume to these types of instances. See [documentation](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateEndpointConfig.html#sagemaker-CreateEndpointConfig-request-KmsKeyId).

As a result, passing any KMS key with such instance is nonsensical, the only admissible value is `None`. The issue is, if the user attempts to pass `None`, the intelligent default can substitute a value, causing the service call to fail and the user has no recourse.

Reply:
>As a result, passing any KMS key with such instance is nonsensical, the only admissible value is None. The issue is, if the user attempts to pass None, the intelligent default can substitute a value, causing the service call to fail and the user has no recourse.

AFAIK, intelligent defaults will take the user supplied value (None) over the ones set in Inteliigent defaults. @rubanh correct me if Im wrong. But with the flexibility we want to provide with PySDK, the user need not figure out that the instance doesn't support kms volume and explicitly pass None for such a case.

Reply:
So actually @JGuinegagne is right about this case @mufaddal-rohawala , if the user supplies None, the Defaults Config understands that as "the user didnt supply a value" (because it doesnt know if the None was a explicitly or implicitly provided)

And then to followup on my original comment, thanks for linking to documentation.

So lets break this down into separate questions.
1. Should the VolumeKMS key be skipped if all instances in a request have local storage when...
     1. ... a user passes in a non-None VolumeKMS key explicitly?
Right now, seems this PR in most cases says NO to this question. But in this one case above it says YES
     1. ... the VolumeKMS key was explicitly set in the DefaultConfig?
The goal of this PR is YES. If we are going with this I think at minimum we need to make sure Admins understand this behavior when they set a VolumeKMS key in the Config.
1. Should the VolumeKMS key be skipped if even one instance in a request supports KMS? NO.